### PR TITLE
agent: Statically link kata-agent to GNU libc

### PR DESCRIPTION
**Pull request by fidencio**
Whenever GNU libc is being used, let's ensure kata-agent gets statically
linked against, as done when using musl.

Turns out that when using rust >= 1.49.0 it's now possible and can be
easily achieved by passing "-C target-feature=+crt-static", as part of
the RUSTFLAGS, to cargo.

Fixes: #1055

@lifupan, @Tim-Zhang, is there some easy way to check that one **must** use 1.49.0 or newer?

@Jakob-Naucke, please, take a look and see if it fits your needs.

### CODE REVIEW

**Comment by fidencio** [comment]
Hmm. I start to think I should also add a new CI using the glibc agent.

**Comment by fidencio** [comment]
@rhafer, it may also be interesting for you.

**Comment by eddyb** [comment]
I'm glad to see the conversion to `capctl` was so straightforward!

**Comment by fidencio** [comment]
> I'm glad to see the conversion to `capctl` was so straightforward!

@eddyb, I must admit I was surprised as well.
Thanks a lot for all the help and support figuring out what was dragging "-Wl,-Bdynamic" "-lc" in.

**Comment by Jakob-Naucke** [comment]
@fidencio hmmm. First of all, a big thank you for your fix and thanks @eddyb for your advice.
On s390x, it's only a partial success -- see this diff of `ldd` outputs for static linkage vs dynamic. (The outputs were sorted and the addresses were removed.)
```
        /lib/ld64.so.1                                                  /lib/ld64.so.1 
        libc.so.6 => /lib/s390x-linux-gnu/libc.so.6                     libc.so.6 => /lib/s390x-linux-gnu/libc.so.6 
                                                              >         libdl.so.2 => /lib/s390x-linux-gnu/libdl.so.2 
                                                              >         libgcc_s.so.1 => /lib/s390x-linux-gnu/libgcc_s.so.1 
                                                              >         libpthread.so.0 => /lib/s390x-linux-gnu/libpthread.so
        libutil.so.1 => /lib/s390x-linux-gnu/libutil.so.1               libutil.so.1 => /lib/s390x-linux-gnu/libutil.so.1 
        linux-vdso64.so.1                                               linux-vdso64.so.1 
```
i.e. some external deps are removed, but the end result is not statically linked -- most notably, my problem child `libc.so.6` is still there.
For comparison, a bare `cargo init` (hello world) has
```
   /lib/ld64.so.1 
   libc.so.6 => /lib/s390x-linux-gnu/libc.so.6 
   libdl.so.2 => /lib/s390x-linux-gnu/libdl.so.2 
   libgcc_s.so.1 => /lib/s390x-linux-gnu/libgcc_s.so.1 
   libpthread.so.0 => /lib/s390x-linux-gnu/libpthread.so.0 
   linux-vdso64.so.1
```
and with static linkage I get `not a dynamic executable`, as expected.
For reference, here's the verbose output of compiling the agent:
```
   Compiling kata-agent v0.1.0 (/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent)
     Running `CARGO=/home/ubuntu/.rustup/toolchains/stable-s390x-unknown-linux-gnu/bin/cargo CARGO_BIN_NAME=kata-agent CARGO_CRATE_NAME=kata_agent CARGO_MANIFEST_DIR=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent CARGO_PKG_AUTHORS='The Kata Containers community <kata-dev@lists.katacontainers.io>' CARGO_PKG_DESCRIPTION='' CARGO_PKG_HOMEPAGE='' CARGO_PKG_LICENSE='' CARGO_PKG_LICENSE_FILE='' CARGO_PKG_NAME=kata-agent CARGO_PKG_REPOSITORY='' CARGO_PKG_VERSION=0.1.0 CARGO_PKG_VERSION_MAJOR=0 CARGO_PKG_VERSION_MINOR=1 CARGO_PKG_VERSION_PATCH=0 CARGO_PKG_VERSION_PRE='' CARGO_PRIMARY_PACKAGE=1 LD_LIBRARY_PATH='/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/release/deps:/home/ubuntu/.rustup/toolchains/stable-s390x-unknown-linux-gnu/lib:/home/ubuntu/.rustup/toolchains/stable-s390x-unknown-linux-gnu/lib:/usr/lib64' rustc --crate-name kata_agent --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 -C lto -C metadata=09145698c4a70bac -C extra-filename=-09145698c4a70bac --out-dir /home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps --target s390x-unknown-linux-gnu -L dependency=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps -L dependency=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/release/deps --extern anyhow=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libanyhow-f576a242b83e8e56.rlib --extern capctl=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libcapctl-55e52502bb8af9df.rlib --extern cgroups=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libcgroups-b3a6b8ff2536bb03.rlib --extern lazy_static=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/liblazy_static-3353a8bccfe23496.rlib --extern libc=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/liblibc-e23eaaca3725ece7.rlib --extern log=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/liblog-55779ec0a4dd39b7.rlib --extern logging=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/liblogging-00c9e5834fef85a6.rlib --extern netlink=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libnetlink-2909c0708714467f.rlib --extern nix=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libnix-9c56b02e20d72422.rlib --extern oci=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/liboci-6a2a50e9d3c1b4b1.rlib --extern procfs=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libprocfs-75eac979ad0b1df0.rlib --extern prometheus=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libprometheus-a1ce4b5b698d76a4.rlib --extern protobuf=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libprotobuf-24cdce28f3f1cb37.rlib --extern protocols=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libprotocols-7dfe9e3644c93536.rlib --extern regex=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libregex-968b859cdaebfc90.rlib --extern rustjail=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/librustjail-8afc6b518781b1fe.rlib --extern scan_fmt=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libscan_fmt-7bab94f0dc9139af.rlib --extern scopeguard=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libscopeguard-f2776bc6c5d21c8f.rlib --extern serde_json=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libserde_json-2a06f412efc5aae5.rlib --extern signal_hook=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libsignal_hook-210a9bf9cd9de359.rlib --extern slog=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libslog-dc67b2f0b586f84c.rlib --extern slog_scope=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libslog_scope-6ce30e433bce3705.rlib --extern slog_stdlog=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libslog_stdlog-31e795bbb4982ccc.rlib --extern tempfile=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libtempfile-271294a402c9f9cf.rlib --extern ttrpc=/home/ubuntu/go/src/github.com/kata-containers/kata-containers/src/agent/target/s390x-unknown-linux-gnu/release/deps/libttrpc-4078e6ee82de2cd8.rlib -C target-feature=+crt-static --deny warnings`
```
@eddyb could this happen due to yet another dependency that leads to dynamic linkage? How would I go about finding out which one it is?

File: src/agent/Makefile [hunk 1]
```diff
@@ -43,6 +43,11 @@ endif
 
 
 EXTRA_RUSTFLAGS :=
+
+ifeq ($(LIBC), gnu)
+    override EXTRA_RUSTFLAGS = -C target-feature=+crt-static
+endif
+
 ifeq ($(ARCH), aarch64)
     override EXTRA_RUSTFLAGS = -C link-arg=-lgcc
```
Comment (Jakob-Naucke) [inline src/agent/Makefile:52]: @fidencio doesn't this mean that if someone were to build with glibc on aarch64, they'd miss out on static linkage? (Not that that's the most common thing, but still) I think you need `EXTRA_RUSTFLAGS += -C link-arg=-lgcc`

Comment (fidencio) [reply src/agent/Makefile:52]: Indeed, let me change this, thanks for the review!

**Comment by jodh-intel** [comment]
@fidencio - Here's how you can a enforce minimum rust version:


```diff
diff --git a/src/agent/Cargo.toml b/src/agent/Cargo.toml
index 07fc2f6b..a3952c94 100644
--- a/src/agent/Cargo.toml
+++ b/src/agent/Cargo.toml
@@ -4,6 +4,9 @@ version = "0.1.0"
 authors = ["The Kata Containers community <kata-dev@lists.katacontainers.io>"]
 edition = "2018"

+[build-dependencies]
+rustc_version = "0.3.0"
+
 [dependencies]
 oci = { path = "oci" }
 logging = { path = "../../pkg/logging" }
diff --git a/src/agent/build.rs b/src/agent/build.rs
new file mode 100644
index 00000000..395e3dca
--- /dev/null
+++ b/src/agent/build.rs
@@ -0,0 +1,28 @@
+// Copyright (c) 2021 ...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+extern crate rustc_version;
+
+use rustc_version::{version, Version};
+use std::io::{self, Write};
+use std::process::exit;
+
+const MIN_VERSION: &str = "1.49.0";
+
+fn main() {
+    let version = version().unwrap();
+    let min_version = Version::parse(MIN_VERSION).unwrap();
+
+    if version < min_version {
+        writeln!(
+            &mut io::stderr(),
+            "ERROR: Need rustc version {} or newer, got {}",
+            MIN_VERSION,
+            version
+        )
+        .unwrap();
+        exit(1);
+    }
+}
```

**Comment by fidencio** [comment]
> @fidencio - Here's how you can a enforce minimum rust version:

....

I've added an extra patch (keeping your authorship, John) with the check. And here's the output whenever trying to build with an older version:
```
error: failed to run custom build command for `kata-agent v0.1.0 (/home/fidencio/go/src/github.com/kata-containers/kata-containers/src/agent)`

Caused by:
  process didn't exit successfully: `/home/fidencio/go/src/github.com/kata-containers/kata-containers/src/agent/target/release/build/kata-agent-07836618ae340e04/build-script-build` (exit code: 1)
  --- stderr
  ERROR: Need rustc version 1.49.0 or newer, got 1.48.0
warning: build failed, waiting for other jobs to finish...
error: build failed
make: *** [Makefile:116: target/x86_64-unknown-linux-gnu/release/kata-agent] Error 101
```

Thanks!

**Comment by fidencio** [comment]
@Amulyam24, @bpradipt, mind to give it a try on ppc64le? I wonder if you'll be facing the same issues as @Jakob-Naucke faced on s390x.

**Comment by fidencio** [comment]
@Jakob-Naucke, just for the sake of completeness:
```
fidencio@kafka:~/go/src/github.com/fidencio/kata-containers/src/agent$ ldd ./target/aarch64-unknown-linux-gnu/release/kata-agent
	linux-vdso.so.1 (0x0000007f8c3df000)
	libutil.so.1 => /lib/aarch64-linux-gnu/libutil.so.1 (0x0000007f8c37a000)
	libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000007f8c221000)
	/lib/ld-linux-aarch64.so.1 (0x0000007f8c3b3000)

```

Yep, aarch64 seems to hit exactly the same issue.

**Comment by fidencio** [comment]
For now I'm enforcing to statically link kata-agent against glibc **only** on x86_64.
And a huge thanks to @Jakob-Naucke for testing and debugging the issue on s390x.

**Review by snir911 — COMMENTED** [review]
nit typo in commit msg, s/isue/issue

**Comment by Amulyam24** [comment]
> @Amulyam24, @bpradipt, mind to give it a try on ppc64le? I wonder if you'll be facing the same issues as @Jakob-Naucke faced on s390x.

@fidencio, on trying the fix, its the same issue on ppc64le too.
```
 ldd ./target/powerpc64le-unknown-linux-gnu/release/kata-agent
	linux-vdso64.so.1 (0x00007fffb1d80000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007fffb1d40000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fffb1b20000)
	/usr/lib/ld.so.1 => /lib64/ld64.so.2 (0x00007fffb1da0000)
```

**Comment by Jakob-Naucke** [comment]
@Amulyam24 interesting. Would you mind trying this:
- remove [this line](https://github.com/rust-lang/libc/blob/master/src/unix/linux_like/linux/gnu/b64/powerpc64/mod.rs#L1049) in a local clone of libc
- patch that into your agent with
```
[patch.crates-io]
libc = { path = "/your/libc/here" }
```
in your `Cargo.toml`
- rebuild and re-check the `ldd` output

**Comment by fidencio** [comment]
Can we have this one reviewed / merged? :-)
@lifupan, @bergwolf, @jodh-intel, @egernst, @Jakob-Naucke

**Review by Jakob-Naucke — APPROVED** [review]
Yes, it makes sense to merge the change for x86 now. LGTM and thank you @fidencio !

**Review by lifupan — APPROVED** [review]

**Comment by lifupan** [comment]
> Yes, it makes sense to merge the change for x86 now. LGTM and thank you @fidencio !

Hi  @Jakob-Naucke 

Did the static link work for your s390?

**Comment by Jakob-Naucke** [comment]
@lifupan not properly. [aarch64](https://github.com/kata-containers/kata-containers/pull/1236#issuecomment-756833916), [ppc64le](https://github.com/kata-containers/kata-containers/pull/1236#issuecomment-757900760), and [s390x](https://github.com/kata-containers/kata-containers/pull/1236#issuecomment-756662752) (see also standalone issue #1239) have dynamic object dependencies left, which seems to have to do with libc (see also my issue https://github.com/rust-lang/libc/issues/2033 over there). I can alter libc to get a statically linked binary, but that change hasn't really been discussed yet.

File: src/agent/build.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    if version < min_version {
+        writeln!(
+            &mut io::stderr(),
+            "ERROR: Need rustc version {} or newer, got {}",
+            MIN_VERSION,
+            version
+        )
+        .unwrap();
+        exit(1);
```
Comment (mxpv) [inline src/agent/build.rs:26]: nit: Can be just
```rust
panic!("ERROR: Need rustc version {} or newer, got {}", MIN_VERSION, version);
```

Comment (fidencio) [reply src/agent/build.rs:26]: Thanks, @mxpv, will do the change!

**Comment by Jakob-Naucke** [comment]
I had originally gotten into this due to #675, where a dynamically linked agent would fail on s390x with an Ubuntu 20.04 container on an Ubuntu 18.04 guest (the default) (on a 20.04 build host). That combination does, btw, work with the statically linked agent.
However, I had continued my investigations because with the statically linked binary, a Fedora 33 container (which was working with dynamic) would yield the same issue, albeit with a stack trace:
<details><summary>log excerpt</summary><p>

```
{"msg":"wait for child process ready to run exec","level":"INFO","ts":"2021-01-12T17:25:56.314671834+00:00","subsystem":"container","name":"kata-agent","version":"0.1.0","pid":"139","module":"rustjail","cid":"hello","source":"agent","action":"join-namespaces","eid":"hello"}
[    0.472218] User process fault: interruption code 0010 ilc:2 in libc-2.32.so[3ffa2e80000+1c4000]
[    0.472303] Failing address: 0000000000000000 TEID: 0000000000000800
[    0.472342] Fault in primary space mode while using user ASCE.
[    0.472375] AS:000000007e2081c7 R3:000000007e390007 S:0000000000000020 
[    0.472424] CPU: 0 PID: 157 Comm: kata-agent Not tainted 5.4.71 #1
[    0.472467] Hardware name: IBM 8561 T01 701 (KVM/Linux)
[    0.472501] User PSW : 0705100180000000 000003ffa2fcdf86
[    0.472537]            R:0 T:1 IO:1 EX:1 Key:0 M:1 W:0 P:1 AS:0 CC:1 PM:0 RI:0 EA:3
[    0.472586] User GPRS: 000000000000000a 00000000000000e5 fffffffffffffcc0 0000000001b0b960
[    0.472650]            000000000000001e 0000000000000072 0000000000000000 000003ffa3047d20
[    0.472698]            000003ffa2d00330 0000000001b0bd5f 000003fffa5faea8 0000000001b0b960
[    0.472742]            00000000019beaf8 0000000000000400 000003ffa2fcdf48 000003fffa5fad50
[    0.472792] User Code: 000003ffa2fcdf78: b904003bttlgrt%r3,%r11
[    0.472792]            000003ffa2fcdf7c: eb110001000dtsllgt%r1,%r1,1
[    0.472792]           #000003ffa2fcdf82: 41116001ttlat%r1,1(%r1,%r6)
[    0.472792]           >000003ffa2fcdf86: 91201000tttmt0(%r1),32
[    0.472792]            000003ffa2fcdf8a: a7840011ttbrct8,000003ffa2fcdfac
[    0.472792]            000003ffa2fcdf8e: e31030010090tllgct%r1,1(%r3)
[    0.472792]            000003ffa2fcdf94: 41303001ttlat%r3,1(%r3)
[    0.472792]            000003ffa2fcdf98: 1851ttlrt%r5,%r1
[    0.473077] Last Breaking-Event-Address:
[    0.473099]  [<0000000000000001>] 0x1
{"msg":"create container process error process: 139 failed to receive sync message from peer: got msg length: 0, expected: 4","level":"ERRO","ts":"2021-01-12T17:25:56.357285403+00:00","module":"rustjail","source":"agent","name":"kata-agent","pid":"139","cid":"hello","eid":"hello","version":"0.1.0","subsystem":"container"}
{"msg":"child exited unexpectedly","level":"INFO","ts":"2021-01-12T17:25:56.357431731+00:00","name":"kata-agent","pid":"139","child-pid":"156","subsystem":"signals","version":"0.1.0","source":"agent"}
```
</details>

(happens on both default Ubuntu 18.04 and Fedora 30 guests, notice `libc-2.32.so`, which is only used by Fedora 33 here)
I found that **this is reproducible on x86_64**, i.e. certain combinations of guest and container libc will fail even (or perhaps especially) with static linkage. I'm assuming that `libc.so.6` is still pulled in somehow.
@fidencio DYT that there is a way around this? If not, how should we handle incompatible combinations? An obvious answer would be to keep guest images up to date, but ideally, it shouldn't be necessary having to maintain them this much. (Maintenance can go beyond "set a new version string as the default", e.g. for Ubuntu, it doesn't seem possible to have chrony and systemd within one system any more in 20.04 due to systemd-timesync.)

**Comment by fidencio** [comment]
I'm really sorry for the disruption but in order to do the branch name changes announced [here](http://lists.katacontainers.io/pipermail/kata-dev/2021-January/001695.html) I'm adding the "do-not-merge" and the "needs-branch-adjustment" labels to the PR.

I'll add a new comment once the branch changes are done and once it happens I'd like to ask you to resubmit the PR against the new `main` branch, possibly linking to this PR in case you find the conversations / reviews already made important to proceed.

**Comment by fidencio** [comment]
Kata Containers repos are back to business.  Thanks a lot for your patience and I'm very happy to announce that there's no changes needed from your side as renaming the branches automagically reflected here.

If you notice something weird, some missing CI, please, do not hesitate on contacting us here or on #kata-dev on bit.ly/katacontainersslack.

Also, this is a good opportunity to have your PR rebased atop of `main`, to ensure your PR follow all the guidelines (https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format) enforced by this community, or even to re-evaluate whether the PR is currently needed.

Last but not least,  as the default branch has been renamed to `main`, if you have a local clone you'd need to update it and you do so by running:
```
git branch -m 2.0-dev main
git fetch origin
git branch -u origin/main main
```

Happy hacking!

**Comment by jodh-intel** [comment]
Hi @fidencio - could you rebase as this branch is now conflicted?

File: src/agent/build.rs (hunk 2 above)
Comment (fidencio) [reply src/agent/build.rs]: Suggestion taken, thanks @mxpv!

**Comment by fidencio** [comment]
@jodh-intel, do you think it'd be benifitial to have a CI running using the LIBC=gnu to merge together with these changes?

**Comment by clnperez** [comment]
@fidencio FYI -- there's a typo in the summary here that made me crazy all morning today. /s/ctr-static/crt-static  

I copy-pasted that and couldn't for the life of me figure out why it wasn't working any more when it had earlier in the week :D

**Comment by fidencio** [comment]
Adding a dnm label till I investigate some issues raised by @Jakob-Naucke.

**Comment by Jakob-Naucke** [comment]
@fidencio since libc has the fix for non-x86 static linkage in its [new release 0.2.86](https://github.com/rust-lang/libc/releases/tag/0.2.86), I suppose you could update to that version and remove the limitation to x86. This would also close #1239.

**Comment by fidencio** [comment]
@mxpv, @snir911, @lifupan,

I've a conversation with @Jakob-Naucke Today, trying to reproduce the issues he faced both on s390x and x86_64.  While I was able to reproduce one of the issues (https://github.com/kata-containers/kata-containers/issues/675), I wasn't able to reproduce the glibc issue he had when using the agent built on ubuntu 20.04, inside an ubuntu 18.04 roots, spawning a fedora 33 container.

In any case, I feel that statically linking to glibc is, although desired, still fragile and I've reworked the patches adding yet another variable that can be passed to make: `GNU_STATIC=yes|no`, which has "no"  as its default value.

So, my suggestion is:
* Let's have this one merged, but not enabled by default;
* Whoever wants to play & experiment with this, just have fun :-)
* To build the agent in a way you take advantage of this work, do: `make LIBC=gnu GNU_STATIC=yes`
* This will bump the minimum required rust version to v1.49.0;
  * Is there a way to do so only when building using `LIBC=gnu GNU_STATIC=yes`? Not sure, but I'd be fine doing that if there's an easy way to do so;

Please, the ones who reviewed this series, take a second look whenever it's possible.

File: src/agent/Makefile [hunk 3]
```diff
@@ -53,6 +55,16 @@ ifeq ($(ARCH), aarch64)
     $(warning "WARNING: aarch64-musl needs extra symbols from libgcc")
 endif
 
+# The only architecture we can currently link the kata-agent
+# against glibc statically is x86_64.
+# Please, for more information refer to:
+# https://github.com/kata-containers/kata-containers/issues/1239
```
Comment (Jakob-Naucke) [inline src/agent/Makefile:61]: @fidencio I have tested on s390x, this can be removed :)

Comment (fidencio) [reply src/agent/Makefile]: Oh yeah, I've removed the condition but not the comment. Removed now, thanks!

**Comment by cmaf** [comment]
Hi @fidencio, could you please rebase?

**Comment by fidencio** [comment]
Adding the wip so I can revisit this later this week.

**Comment by fidencio** [comment]
I'm closing this one based on the discussion we had in the architecture commitee meeting.
I'll, however, open a different PR doing the needed / desired version bumps.

**Pull request closed without merging**